File: scala/src/main/java/com/cjl_magistri/tinyweb/TinyWeb.java
package com.cjl_magistri.tinyweb;

import java.util.List;
import java.util.Map;

public class TinyWeb {
    private final Map<String, Controller> routes;
    private final List<Filter> filters;

    public TinyWeb(Map<String, Controller> routes, List<Filter> filters) {
        this.routes = routes;
        this.filters = filters;
    }

    public HttpResponse handleRequest(HttpRequest request) {
        HttpRequest currentRequest = request;
        for (Filter toApply: filters) {
            currentRequest = toApply.doFilter(currentRequest);
        }

        Controller controller = routes.get(currentRequest.getPath());
        if (null == controller) {
            return null;
        }
        return controller.handleRequest(currentRequest);
    }

}

File: scala/src/main/scala/com/cjl_magistri/tinyweb/TinyWeb.scala
package com.cjl_magistri.tinyweb

class TinyWeb(routeMap: Map[String, Controller], filters: List[(HttpRequest) => HttpRequest]) {
  def handleRequest(httpRequest: HttpRequest): Option[HttpResponse] = {
    // Compose all the filters together into a single, composed filter.
    val composedFilter = filters.reverse.reduceLeft((composed, next) => composed compose next)

    // Apply all the filters to the request.
    val filteredRequest = composedFilter(httpRequest)

    // Calculate the controller (option) that handles the request and ...
    val controller = routeMap.get(filteredRequest.path)

    // Map that controller option to a request by invoking handleRequest of the controller.
    controller map { controller => controller.handleRequest(filteredRequest) }
  }
}